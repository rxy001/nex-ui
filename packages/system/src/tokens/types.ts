import type { RawCSSProperties, Dictionary } from '../types'
import type { Token } from './createToken'

export type TokenValue = string | number

export type TokenCategories =
  | 'fontFamilies'
  | 'fontSizes'
  | 'fontWeights'
  | 'lineHeights'
  | 'colors'
  | 'spacing'
  | 'sizes'
  | 'borders'
  | 'radii'

export type ColorsDefinition = Dictionary<
  | {
      50?: RawCSSProperties['color']
      100?: RawCSSProperties['color']
      200?: RawCSSProperties['color']
      300?: RawCSSProperties['color']
      400?: RawCSSProperties['color']
      500?: RawCSSProperties['color']
      600?: RawCSSProperties['color']
      700?: RawCSSProperties['color']
      800?: RawCSSProperties['color']
      900?: RawCSSProperties['color']
      contrastText?: RawCSSProperties['color']
    }
  | RawCSSProperties['color']
>

export type RadiiDefinition = Dictionary<string | number>

export type SpacingDefinition = Dictionary<string | number>

export type SizesDefinition = Dictionary<string | number>

export type FontSizesDefinition = Dictionary<string | number>

export type FontWeightsDefinition = Dictionary<string | number>

export type LineHeightsDefinition = Dictionary<string | number>

export type BordersDefinition = Dictionary<string | number>

export type FontFamiliesDefinition = Dictionary<string>

type ResponsiveColor = {
  _DEFAULT?: RawCSSProperties['color']
  _light?: RawCSSProperties['color']
  _dark?: RawCSSProperties['color']
}
type ExclusiveColor =
  | {
      DEFAULT?: RawCSSProperties['color'] | ResponsiveColor
      _DEFAULT?: never
      _light?: never
      _dark?: never
    }
  | ({
      DEFAULT?: never
    } & ResponsiveColor)

type NestedColor =
  | RawCSSProperties['color'] // Directly a color value
  | (ExclusiveColor & {
      [key: string]: NestedColor // Nested colors, applying the mutually exclusive rule
    }) // Either `DEFAULT` or ResponsiveColor properties

type SemanticTokenDefinition<T> =
  | T
  | ({
      DEFAULT?: T
    } & { [key: string]: SemanticTokenDefinition<T> })

export type SemanticColorDefinition = Dictionary<NestedColor>

export type SemanticSpacingDefinition = Dictionary<
  SemanticTokenDefinition<string | number>
>

export type SemanticSizesDefinition = Dictionary<
  SemanticTokenDefinition<string | number>
>

export type SemanticFontFamiliesDefinition = Dictionary<
  SemanticTokenDefinition<string>
>

export type SemanticFontSizesDefinition = Dictionary<
  SemanticTokenDefinition<string | number>
>

export type SemanticFontWeightsDefinition = Dictionary<
  SemanticTokenDefinition<string | number>
>

export type SemanticLineHeightsDefinition = Dictionary<
  SemanticTokenDefinition<string | number>
>
export type SemanticBordersDefinition = Dictionary<
  SemanticTokenDefinition<string | number>
>
export type SemanticRadiiDefinition = Dictionary<
  SemanticTokenDefinition<string | number>
>

export type SemanticTokensDefinition = {
  colors?: SemanticColorDefinition
  spacing?: SemanticSpacingDefinition
  sizes?: SemanticSizesDefinition
  fontFamilies?: SemanticFontFamiliesDefinition
  fontSizes?: SemanticFontSizesDefinition
  fontWeights?: SemanticFontWeightsDefinition
  lineHeights?: SemanticLineHeightsDefinition
  borders?: SemanticBordersDefinition
  radii?: SemanticRadiiDefinition
}

export type TokensDefinition = {
  colors?: ColorsDefinition
  spacing?: SpacingDefinition
  sizes?: SizesDefinition
  fontFamilies?: FontFamiliesDefinition
  fontSizes?: FontSizesDefinition
  fontWeights?: FontWeightsDefinition
  lineHeights?: LineHeightsDefinition
  borders?: BordersDefinition
  radii?: RadiiDefinition
}

export type CreateTokensConfig = {
  tokens: TokensDefinition
  semanticTokens: SemanticTokensDefinition
  prefix: string
}

export type TokenMap = Map<string, Token>

export type CssVarMap = Map<string, Map<string, string | number>>
