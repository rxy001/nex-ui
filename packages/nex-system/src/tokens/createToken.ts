import type { TokenValue, TokenCategory } from './types'

export class Token {
  path: string[]

  value: TokenValue

  name: string

  originalValue: TokenValue

  category: TokenCategory

  derivative: boolean

  cssVar?: {
    var: string
    ref: string
  }

  constructor({
    path,
    value,
    category,
    name,
    originalValue,
    derivative,
    cssVar,
  }: Token) {
    this.path = path
    this.value = value
    this.category = category
    this.name = name
    this.originalValue = originalValue
    this.derivative = derivative
    this.cssVar = cssVar
  }
}

export function createToken(config: Token) {
  return new Token(config)
}
