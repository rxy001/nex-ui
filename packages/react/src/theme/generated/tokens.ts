export interface Tokens {
  colors:
    | 'transparent'
    | 'current'
    | 'white'
    | 'black'
    | 'blue.50'
    | 'blue.100'
    | 'blue.200'
    | 'blue.300'
    | 'blue.400'
    | 'blue.500'
    | 'blue.600'
    | 'blue.700'
    | 'blue.800'
    | 'blue.900'
    | 'blue.contrastText'
    | 'gray.50'
    | 'gray.100'
    | 'gray.200'
    | 'gray.300'
    | 'gray.400'
    | 'gray.500'
    | 'gray.600'
    | 'gray.700'
    | 'gray.800'
    | 'gray.900'
    | 'gray.contrastText'
    | 'pink.50'
    | 'pink.100'
    | 'pink.200'
    | 'pink.300'
    | 'pink.400'
    | 'pink.500'
    | 'pink.600'
    | 'pink.700'
    | 'pink.800'
    | 'pink.900'
    | 'pink.contrastText'
    | 'purple.50'
    | 'purple.100'
    | 'purple.200'
    | 'purple.300'
    | 'purple.400'
    | 'purple.500'
    | 'purple.600'
    | 'purple.700'
    | 'purple.800'
    | 'purple.900'
    | 'purple.contrastText'
    | 'cyan.50'
    | 'cyan.100'
    | 'cyan.200'
    | 'cyan.300'
    | 'cyan.400'
    | 'cyan.500'
    | 'cyan.600'
    | 'cyan.700'
    | 'cyan.800'
    | 'cyan.900'
    | 'cyan.contrastText'
    | 'yellow.50'
    | 'yellow.100'
    | 'yellow.200'
    | 'yellow.300'
    | 'yellow.400'
    | 'yellow.500'
    | 'yellow.600'
    | 'yellow.700'
    | 'yellow.800'
    | 'yellow.900'
    | 'yellow.contrastText'
    | 'orange.50'
    | 'orange.100'
    | 'orange.200'
    | 'orange.300'
    | 'orange.400'
    | 'orange.500'
    | 'orange.600'
    | 'orange.700'
    | 'orange.800'
    | 'orange.900'
    | 'orange.contrastText'
    | 'red.50'
    | 'red.100'
    | 'red.200'
    | 'red.300'
    | 'red.400'
    | 'red.500'
    | 'red.600'
    | 'red.700'
    | 'red.800'
    | 'red.900'
    | 'red.contrastText'
    | 'green.50'
    | 'green.100'
    | 'green.200'
    | 'green.300'
    | 'green.400'
    | 'green.500'
    | 'green.600'
    | 'green.700'
    | 'green.800'
    | 'green.900'
    | 'green.contrastText'
  sizes:
    | '1'
    | '2'
    | '3'
    | '4'
    | '5'
    | '6'
    | '7'
    | '8'
    | '9'
    | '10'
    | 'max'
    | 'min'
    | 'full'
    | 'px'
  spacing:
    | '1'
    | '2'
    | '3'
    | '4'
    | '5'
    | '6'
    | '7'
    | '8'
    | 'px'
    | '0.5'
    | '1.5'
    | '2.5'
    | '3.5'
  fontFamilies: 'heading' | 'body' | 'mono'
  fontSizes: 'sm' | 'md' | 'lg'
  fontWeights:
    | 'hairline'
    | 'thin'
    | 'light'
    | 'normal'
    | 'medium'
    | 'semibold'
    | 'bold'
    | 'extrabold'
    | 'black'
  borders: 'xs' | 'sm' | 'md' | 'lg' | 'xl'
  radii: 'sm' | 'md' | 'lg'
  lineHeights: 'normal' | 'none' | 'shorter' | 'short' | 'base' | 'tall'
}

export interface Colors {
  transparent?: 'transparent'
  current?: 'currentColor'
  white?: '#fff'
  black?: '#000'
  blue?: {
    50?: '#e6f4ff'
    100?: '#dbeafe'
    200?: '#bfdbfe'
    300?: '#a3cfff'
    400?: '#60a5fa'
    500?: '#3b82f6'
    600?: '#2563eb'
    700?: '#173da6'
    800?: '#1a3478'
    900?: '#14204a'
    contrastText?: '#fff'
  }
  gray?: {
    50?: '#fafafa'
    100?: '#f4f4f5'
    200?: '#e4e4e7'
    300?: '#d4d4d8'
    400?: '#a1a1aa'
    500?: '#71717a'
    600?: '#52525b'
    700?: '#3f3f46'
    800?: '#27272a'
    900?: '#18181b'
    contrastText?: '#fff'
  }
  pink?: {
    50?: '#fdf2f8'
    100?: '#fce7f3'
    200?: '#fbcfe8'
    300?: '#f9a8d4'
    400?: '#f472b6'
    500?: '#ec4899'
    600?: '#db2777'
    700?: '#a41752'
    800?: '#6d0e34'
    900?: '#45061f'
    contrastText?: '#fff'
  }
  purple?: {
    50?: '#faf5ff'
    100?: '#f3e8ff'
    200?: '#e9d5ff'
    300?: '#d8b4fe'
    400?: '#c084fc'
    500?: '#a855f7'
    600?: '#9333ea'
    700?: '#641ba3'
    800?: '#4a1772'
    900?: '#2f0553'
    contrastText?: '#fff'
  }
  cyan?: {
    50?: '#ecfeff'
    100?: '#cffafe'
    200?: '#a5f3fc'
    300?: '#67e8f9'
    400?: '#22d3ee'
    500?: '#06b6d4'
    600?: '#0891b2'
    700?: '#0c5c72'
    800?: '#134152'
    900?: '#072a38'
    contrastText?: '#fff'
  }
  yellow?: {
    50?: '#fefce8'
    100?: '#fef9c3'
    200?: '#fef08a'
    300?: '#fde047'
    400?: '#facc15'
    500?: '#eab308'
    600?: '#ca8a04'
    700?: '#a16207'
    800?: '#854d0e'
    900?: '#713f12'
    contrastText?: '#fff'
  }
  orange?: {
    50?: '#fff7ed'
    100?: '#ffedd5'
    200?: '#fed7aa'
    300?: '#fdba74'
    400?: '#fb923c'
    500?: '#f97316'
    600?: '#ea580c'
    700?: '#c2410c'
    800?: '#9a3412'
    900?: '#7c2d12'
    contrastText?: '#fff'
  }
  red?: {
    50?: '#fef2f2'
    100?: '#fee2e2'
    200?: '#fecaca'
    300?: '#fca5a5'
    400?: '#f87171'
    500?: '#ef4444'
    600?: '#dc2626'
    700?: '#b91c1c'
    800?: '#991b1b'
    900?: '#7f1d1d'
    contrastText?: '#fff'
  }
  green?: {
    50?: '#f0fdf4'
    100?: '#dcfce7'
    200?: '#bbf7d0'
    300?: '#86efac'
    400?: '#4ade80'
    500?: '#22c55e'
    600?: '#16a34a'
    700?: '#15803d'
    800?: '#166534'
    900?: '#14532d'
    contrastText?: '#fff'
  }
}

export interface Sizes {
  1?: '4px'
  2?: '8px'
  3?: '12px'
  4?: '16px'
  5?: '20px'
  6?: '24px'
  7?: '28px'
  8?: '32px'
  9?: '36px'
  10?: '40px'
  max?: 'max-content'
  min?: 'min-content'
  full?: '100%'
  px?: '1px'
}

export interface Spacing {
  1?: '4px'
  2?: '8px'
  3?: '12px'
  4?: '16px'
  5?: '20px'
  6?: '24px'
  7?: '28px'
  8?: '32px'
  px?: '1px'
  0.5?: '2px'
  1.5?: '6px'
  2.5?: '10px'
  3.5?: '14px'
}

export interface FontFamilies {
  heading?: 'Roboto, Helvetica, Arial, sans-serif'
  body?: 'Roboto, Helvetica, Arial, sans-serif'
  mono?: 'SFMono-Regular, Menlo,Monaco, Consolas, "Liberation Mono", "Courier New", monospace'
}

export interface FontSizes {
  sm?: '12px'
  md?: '14px'
  lg?: '16px'
}

export interface FontWeights {
  hairline?: '100'
  thin?: '200'
  light?: '300'
  normal?: '400'
  medium?: '500'
  semibold?: '600'
  bold?: '700'
  extrabold?: '800'
  black?: '900'
}

export interface Borders {
  xs?: '0.5px'
  sm?: '1px'
  md?: '2px'
  lg?: '4px'
  xl?: '8px'
}

export interface Radii {
  sm?: '4px'
  md?: '6px'
  lg?: '8px'
}

export interface LineHeights {
  normal?: 'normal'
  none?: '1'
  shorter?: '1.25'
  short?: '1.375'
  base?: '1.5'
  tall?: '1.625'
}
