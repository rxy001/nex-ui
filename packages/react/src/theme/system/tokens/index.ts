import { defineTokens } from '@nex-ui/system'
import { colors } from './colors'
import { spacing } from './spacing'
import { fontFamilies, fontSizes, fontWeights, lineHeights } from './typography'
import { sizes } from './sizes'
import { borders } from './borders'
import { radii } from './radii'
import { borderWidths } from './borderWidths'
import { shadows } from './shadows'
import { transitions } from './transitions'
import { zIndexes } from './zIndexes'

export const tokens = defineTokens({
  colors,
  sizes,
  spacing,
  fontFamilies,
  fontSizes,
  fontWeights,
  borders,
  radii,
  lineHeights,
  borderWidths,
  shadows,
  transitions,
  zIndexes,
})
