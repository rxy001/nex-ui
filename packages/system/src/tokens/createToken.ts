import type { TokenValue, TokenCategory, SemanticTokenValue } from './types'

export class Token {
  path: string[]

  value: string

  name: string

  originalValue: TokenValue | SemanticTokenValue

  category: TokenCategory

  cssVar?: {
    var: string
    ref: string
  }

  conditions?: {
    base?: TokenValue
    light?: TokenValue
    dark?: TokenValue
  }

  constructor({
    path,
    value,
    category,
    name,
    originalValue,
    cssVar,
    conditions,
  }: Token) {
    this.path = path
    this.value = value
    this.category = category
    this.name = name
    this.originalValue = originalValue
    this.cssVar = cssVar
    this.conditions = conditions
  }
}

export function createToken(config: Token) {
  return new Token(config)
}
