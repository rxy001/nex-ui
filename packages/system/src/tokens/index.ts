export { createTokens } from './createTokens'

export type { Token } from './createToken'

export type {
  ColorsDefinition,
  SpacingDefinition,
  SizesDefinition,
  FontFamiliesDefinition,
  FontSizesDefinition,
  FontWeightsDefinition,
  LineHeightsDefinition,
  TokenCategory,
  TokenDefinitions,
  BordersDefinition,
  RadiiDefinition,
  // SemanticDefinition,
} from './types'
