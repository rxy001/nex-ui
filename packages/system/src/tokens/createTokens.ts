import { forEach, isString, walkObject } from '@nex-ui/utils'
import type {
  CreateTokensConfig,
  TokenMap,
  CssVarMap,
  TokenCategories,
} from './types'
import { negate } from '../calc'
import { createToken } from './createToken'
import {
  createCssVarName,
  pathToName,
  checkTokenValue,
  checkTokenCategory,
  isResponsiveColor,
} from '../utils'
import type { Token } from './createToken'

function filterDefault(path: string[]) {
  if (path[0] === 'DEFAULT') return path
  return path.filter((item) => item !== 'DEFAULT')
}

export function createTokens(config: CreateTokensConfig) {
  const { tokens, semanticTokens, prefix } = config

  const tokenMap: TokenMap = new Map()

  const cssVarMap: CssVarMap = new Map()

  let workInProgress: Token | null = null

  function addToTokenMap() {
    if (workInProgress) {
      tokenMap.set(workInProgress.name, workInProgress)
    }
  }

  function createCssVar() {
    if (workInProgress) {
      const { path } = workInProgress

      const variableName = createCssVarName(prefix, path)

      const newValue = `var(${variableName})`

      workInProgress.cssVar = {
        var: variableName,
        ref: newValue,
      }

      workInProgress.value = newValue
    }
  }

  function addNegativeToken() {
    if (workInProgress) {
      const { category, value, path, originalValue, cssVar } = workInProgress
      const newPath = [...path]
      newPath.push(`-${newPath.pop()}`)

      workInProgress = createToken({
        category,
        originalValue,
        path: newPath,
        name: pathToName(newPath),
        value: negate(cssVar?.ref ?? value),
      })

      addToTokenMap()
    }
  }

  function buildCssVars() {
    if (workInProgress) {
      const { conditions, cssVar } = workInProgress

      const variableName = cssVar?.var as string

      forEach(
        conditions,
        (value: string | number | undefined, condition: string) => {
          if (!value) {
            return
          }

          if (!cssVarMap.has(condition)) {
            cssVarMap.set(condition, new Map())
          }

          cssVarMap.get(condition)!.set(variableName, value)
        },
      )
    }
  }

  function handleToken() {
    if (workInProgress) {
      createCssVar()
      buildCssVars()
      addToTokenMap()

      const { category, value } = workInProgress

      switch (category) {
        case 'spacing':
          if (value === '0rem' || value === 0) {
            return
          }
          addNegativeToken()
          break
        default:
          break
      }

      workInProgress = null
    }
  }

  function getTokenValueByCategory(category: TokenCategories) {
    return (value: string) => {
      if (isString(value)) {
        const regex = /\{(.*?)\}/
        const match = value.match(regex)
        if (match && match[1]) {
          const token = tokenMap.get(`${category}.${match[1]}`)

          return token?.value ?? value
        }
      }

      return value
    }
  }

  function workloop() {
    walkObject(
      tokens,
      (value: string | number, path: string[]) => {
        if (!checkTokenValue(value, path)) return

        const category = path[0]

        if (!checkTokenCategory(category)) {
          return
        }

        workInProgress = createToken({
          path,
          value,
          category: category as TokenCategories,
          name: pathToName(path),
          originalValue: value,
          conditions: {
            base: value,
          },
        })

        handleToken()
      },
      {
        predicate: (_: any, path: string[]) => {
          const category = path[0]
          switch (category) {
            case 'colors':
              return path.length > 3
            default:
              return path.length > 2
          }
        },
      },
    )

    walkObject(
      semanticTokens,
      (value: any, path: string[]) => {
        if (!checkTokenCategory(path[0])) {
          return
        }

        const category = path[0] as TokenCategories

        const newPath = filterDefault(path)

        const t = getTokenValueByCategory(category)

        const conditions = isResponsiveColor(value)
          ? {
              base: t(value._DEFAULT),
              dark: t(value._dark),
              light: t(value._light),
            }
          : { base: t(value) }

        workInProgress = createToken({
          value,
          category,
          conditions,
          path: newPath,
          name: pathToName(newPath),
          originalValue: value,
        })

        handleToken()
      },
      {
        predicate: isResponsiveColor,
      },
    )
  }

  workloop()

  return {
    getToken: (key: string) => tokenMap.get(key),
    getGlobalCssVars: () => cssVarMap,
  }
}

export type Tokens = ReturnType<typeof createTokens>
