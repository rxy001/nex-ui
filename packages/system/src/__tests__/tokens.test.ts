import { createTokens } from '../tokens'
import { Token } from '../tokens/createToken'

const PREFIX = 'test'

describe('createTokens', () => {
  it('should work correctly', () => {
    const { getToken, getGlobalCssVars } = createTokens({
      tokens: {
        colors: {
          white: '#fff',
          blue: {
            100: '#cce3fd',
            200: '#99c7fb',
          },
        },
        borderWidths: {
          sm: '1px',
        },
        borders: {
          sm: '1px solid',
        },
        fontFamilies: {
          body: 'system',
        },
        fontSizes: {
          sm: '12px',
        },
        fontWeights: {
          blod: '600',
        },
        lineHeights: {
          base: '1.4',
        },
        radii: {
          sm: '4px',
        },
        shadows: {
          sm: 'shadow',
        },
        sizes: {
          sm: '4px',
        },
        spaces: {
          sm: '4px',
        },
        transitions: {
          all: 'all 0.2s',
        },
        zIndexes: {
          min: 0,
        },
      },
      semanticTokens: {
        colors: {
          primary: {
            _DEFAULT: 'pink',
            _dark: 'green',
            _light: 'yellow',
          },
        },
      },
      prefix: PREFIX,
    })

    expect(getToken('colors.white')).toMatchSnapshot()
    expect(getToken('colors.blue.100')).toMatchSnapshot()
    expect(getGlobalCssVars()).toMatchSnapshot()
  })

  it('should warn when passing invalid tokens', () => {
    const consoleSpy = jest.spyOn(console, 'error')

    createTokens({
      tokens: {
        // @ts-expect-error
        border: '1px solid',
      },
      semanticTokens: {
        // @ts-expect-error
        border: '1px solid',
      },
    })

    expect(consoleSpy).toHaveBeenCalledTimes(2)
    consoleSpy.mockRestore()
  })

  it('should warn when passing invalid token values', () => {
    const consoleSpy = jest.spyOn(console, 'error')
    createTokens({
      tokens: {
        borderWidths: {
          // @ts-expect-error
          sm: null,
          // @ts-expect-error
          md: undefined,
          // @ts-expect-error
          lg: () => {},
          // @ts-expect-error
          xxl: Symbol('xxl'),
        },
      },
      semanticTokens: {
        colors: {
          primary: {
            // @ts-expect-error
            md: null,
            // @ts-expect-error
            xl: () => {},
            // @ts-expect-error
            xxl: Symbol('xxl'),
          },
        },
      },
    })

    expect(consoleSpy).toHaveBeenCalledTimes(7)
    consoleSpy.mockRestore()
  })

  it('should warn when tokens have invalid nested structures', () => {
    const consoleSpy = jest.spyOn(console, 'error')

    const testToken = {
      sm: '1px',
      border: {
        sm: '1px',
      },
    }

    createTokens({
      tokens: {
        colors: {
          black: '#000',
          blue: { 100: 'blue' },
          white: {
            // @ts-expect-error
            metud: {
              100: '#fff',
            },
          },
        },
        // @ts-expect-error
        sizes: testToken,
        // @ts-expect-error
        borderWidths: testToken,
        // @ts-expect-error
        borders: testToken,
        // @ts-expect-error
        fontFamilies: testToken,
        // @ts-expect-error
        fontSizes: testToken,
        // @ts-expect-error
        fontWeights: testToken,
        // @ts-expect-error
        lineHeights: testToken,
        // @ts-expect-error
        radii: testToken,
        // @ts-expect-error
        shadows: testToken,
        // @ts-expect-error
        spaces: testToken,
        // @ts-expect-error
        transitions: testToken,
        // @ts-expect-error
        zIndexes: testToken,
      },
      semanticTokens: {},
      prefix: PREFIX,
    })

    expect(consoleSpy).toHaveBeenCalled()
    expect(consoleSpy).toHaveBeenCalledTimes(13)

    consoleSpy.mockRestore()
  })

  it('should support token reference syntax', () => {
    const errorSpy = jest.spyOn(console, 'error')
    const { getToken, getGlobalCssVars } = createTokens({
      tokens: {
        colors: {
          blue: {
            100: 'blue',
          },
        },
      },
      semanticTokens: {
        colors: {
          primary: '{colors.blue.100}',
          // error case
          secondary: '{colors.blue.900}',
        },
        borders: {
          sm: '1px solid {colors.blue.100}',
        },
      },
      prefix: PREFIX,
    })

    expect(getToken('colors.primary')).toMatchSnapshot()
    expect(errorSpy).toHaveBeenCalled()
    expect(getGlobalCssVars()).toMatchSnapshot()
  })

  it('should support nested semantic tokens', () => {
    const { getToken } = createTokens({
      tokens: {},
      semanticTokens: {
        colors: {
          primary: {
            button: {
              hover: 'yellow',
            },
            input: {
              hover: 'yellowgreen',
            },
          },
          secondary: {
            button: {
              hover: 'wheat',
            },
          },
        },
      },
      prefix: PREFIX,
    })

    expect(getToken('colors.primary.button.hover')?.originalValue).toBe(
      'yellow',
    )

    expect(getToken('colors.primary.input.hover')?.originalValue).toBe(
      'yellowgreen',
    )

    expect(getToken('colors.secondary.button.hover')?.originalValue).toBe(
      'wheat',
    )
  })

  it('should support light and dark conditions', () => {
    const { getToken, getGlobalCssVars } = createTokens({
      tokens: {
        colors: {
          yellow: {
            100: 'yellowgreen',
            200: 'whitesmoke',
          },
        },
      },
      semanticTokens: {
        colors: {
          primary: {
            DEFAULT: {
              _DEFAULT: 'yellow',
              _dark: '{colors.yellow.100}',
              _light: '{colors.yellow.200}',
            },
            hover: {
              _DEFAULT: 'yellow',
              _dark: '{colors.yellow.100}',
              _light: '{colors.yellow.200}',
            },
          },
        },
      },
      prefix: PREFIX,
    })

    expect(getToken('colors.primary')).toMatchSnapshot()
    expect(getToken('colors.primary.hover')).toMatchSnapshot()
    expect(getGlobalCssVars()).toMatchSnapshot()
  })

  it('should define the default value with the DEFAULT key in nested semantic tokens', () => {
    const { getToken } = createTokens({
      tokens: {},
      semanticTokens: {
        colors: {
          primary: {
            button: {
              DEFAULT: 'yellow',
              hover: 'yellowgreen',
            },
          },
        },
      },
      prefix: PREFIX,
    })
    expect(getToken('colors.primary.button')?.originalValue).toBe('yellow')
    expect(getToken('colors.primary.button.hover')?.originalValue).toBe(
      'yellowgreen',
    )
  })
})

describe('Token', () => {
  it('should create a token instance', () => {
    const token = new Token({
      path: ['colors', 'blue', '100'],
      value: 'blue',
      category: 'colors',
      name: 'blue-100',
      originalValue: 'blue',
      cssVar: {
        var: 'var(--color-blue-100)',
        ref: 'colors.blue.100',
      },
      conditions: {
        base: 'blue',
        light: 'light-blue',
        dark: 'dark-blue',
      },
    })

    expect(token).toMatchSnapshot()
  })
})
