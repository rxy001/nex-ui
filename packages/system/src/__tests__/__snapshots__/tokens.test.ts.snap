// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createTokens should support light and dark conditions 1`] = `
{
  "category": "colors",
  "conditions": {
    "base": "yellow",
    "dark": "var(--test-colors-yellow-100)",
    "light": "var(--test-colors-yellow-200)",
  },
  "cssVar": {
    "ref": "var(--test-colors-primary)",
    "var": "--test-colors-primary",
  },
  "name": "colors.primary",
  "originalValue": {
    "_DEFAULT": "yellow",
    "_dark": "{colors.yellow.100}",
    "_light": "{colors.yellow.200}",
  },
  "path": [
    "colors",
    "primary",
  ],
  "value": "var(--test-colors-primary)",
}
`;

exports[`createTokens should support light and dark conditions 2`] = `
{
  "category": "colors",
  "conditions": {
    "base": "yellow",
    "dark": "var(--test-colors-yellow-100)",
    "light": "var(--test-colors-yellow-200)",
  },
  "cssVar": {
    "ref": "var(--test-colors-primary-hover)",
    "var": "--test-colors-primary-hover",
  },
  "name": "colors.primary.hover",
  "originalValue": {
    "_DEFAULT": "yellow",
    "_dark": "{colors.yellow.100}",
    "_light": "{colors.yellow.200}",
  },
  "path": [
    "colors",
    "primary",
    "hover",
  ],
  "value": "var(--test-colors-primary-hover)",
}
`;

exports[`createTokens should support light and dark conditions 3`] = `
Map {
  "base" => Map {
    "--test-colors-yellow-100" => "yellowgreen",
    "--test-colors-yellow-200" => "whitesmoke",
    "--test-colors-primary" => "yellow",
    "--test-colors-primary-hover" => "yellow",
  },
  "dark" => Map {
    "--test-colors-primary" => "var(--test-colors-yellow-100)",
    "--test-colors-primary-hover" => "var(--test-colors-yellow-100)",
  },
  "light" => Map {
    "--test-colors-primary" => "var(--test-colors-yellow-200)",
    "--test-colors-primary-hover" => "var(--test-colors-yellow-200)",
  },
}
`;

exports[`createTokens should support token reference syntax 1`] = `
{
  "category": "colors",
  "conditions": {
    "base": "var(--test-colors-blue-100)",
  },
  "cssVar": {
    "ref": "var(--test-colors-primary)",
    "var": "--test-colors-primary",
  },
  "name": "colors.primary",
  "originalValue": "{colors.blue.100}",
  "path": [
    "colors",
    "primary",
  ],
  "value": "var(--test-colors-primary)",
}
`;

exports[`createTokens should support token reference syntax 2`] = `
Map {
  "base" => Map {
    "--test-colors-blue-100" => "blue",
    "--test-colors-primary" => "var(--test-colors-blue-100)",
    "--test-colors-secondary" => "colors.blue.900",
    "--test-borders-sm" => "1px solid var(--test-colors-blue-100)",
  },
}
`;

exports[`createTokens should work correctly 1`] = `
{
  "category": "colors",
  "conditions": {
    "base": "#fff",
  },
  "cssVar": {
    "ref": "var(--test-colors-white)",
    "var": "--test-colors-white",
  },
  "name": "colors.white",
  "originalValue": "#fff",
  "path": [
    "colors",
    "white",
  ],
  "value": "var(--test-colors-white)",
}
`;

exports[`createTokens should work correctly 2`] = `
{
  "category": "colors",
  "conditions": {
    "base": "#cce3fd",
  },
  "cssVar": {
    "ref": "var(--test-colors-blue-100)",
    "var": "--test-colors-blue-100",
  },
  "name": "colors.blue.100",
  "originalValue": "#cce3fd",
  "path": [
    "colors",
    "blue",
    "100",
  ],
  "value": "var(--test-colors-blue-100)",
}
`;

exports[`createTokens should work correctly 3`] = `
Map {
  "base" => Map {
    "--test-colors-white" => "#fff",
    "--test-colors-blue-100" => "#cce3fd",
    "--test-colors-blue-200" => "#99c7fb",
    "--test-border-widths-sm" => "1px",
    "--test-borders-sm" => "1px solid",
    "--test-font-families-body" => "system",
    "--test-font-sizes-sm" => "12px",
    "--test-font-weights-blod" => "600",
    "--test-line-heights-base" => "1.4",
    "--test-radii-sm" => "4px",
    "--test-shadows-sm" => "shadow",
    "--test-sizes-sm" => "4px",
    "--test-spaces-sm" => "4px",
    "--test-transitions-all" => "all 0.2s",
    "--test-colors-primary" => "pink",
  },
  "dark" => Map {
    "--test-colors-primary" => "green",
  },
  "light" => Map {
    "--test-colors-primary" => "yellow",
  },
}
`;
